src/rust/lib/graph-editor/src/component/visualization/class.rs

```diff
@@ -154,3 +176,126 @@ impl Visualization {
+/// instantiate `JsRenderer`, and the fairly generic `NativeConstructorClass`, that only requires
+/// a function that can create a InstantiationResult. The later can be used as a thin wrapper around
+/// the constructor methods of native visualizations.
+///
+/// Example
+/// --------
+/// ```no_run
+/// use graph_editor::component::visualization;
+/// use graph_editor::component::visualization::Visualization;
+/// use graph_editor::component::visualization::renderer::example::native::BubbleChart;
+/// use ensogl::display::Scene;
+/// use std::rc::Rc;
+///
+/// // Create a `visualization::Class` from a JS source code snippet.
+/// let js_source_class = visualization::JsSourceClass::from_js_source_raw(r#"
+///
+///    class BubbleVisualization {
+///         static inputTypes = ["[[float;3]]"]
+///         onDataReceived(root, data) {}
+///         setSize(root, size) {}
+///     }
+///
+///     return BubbleVisualization;
+///
+/// "#.into());
+///
+/// // Create a `visualization::Class` that instantiates a `BubbleChart`.
+/// let native_bubble_vis_class = visualization::NativeConstructorClass::new(
+///     visualization::ClassAttributes {
+///         name        : "Bubble Visualization (native)".to_string(),
+///         input_types : vec!["[[float;3]]".to_string().into()],
+///     },
+///     |scene:&Scene| Ok(Visualization::new(BubbleChart::new(scene)))
+/// );
+/// ```
+pub trait Class: Debug {
+    /// Provides additional information about the `Class`, for example, which `DataType`s can be
+    /// rendered by the instantiated visualization.
+    fn attributes(&self) -> &ClassAttributes;
+    /// Create new visualization, that is initialised for the given scene. This can fail if the
+    /// `visualization::Class` contains invalid data, for example, JS code that fails to execute,
+    /// or if the scene is in an invalid state.
+    // TODO consider not allowing failing here and require the checking on instantiation of the `Class`.
+    // TODO consider not providing the scene here, but hooking the the shapes/dom elements into the
+    // scene externally.
+    fn instantiate(&self, scene:&Scene) -> InstantiationResult;
+}
+
+#[derive(Clone,Debug,Default)]
+#[allow(missing_docs)]
+pub struct ClassHandle {
+    class : Option<Rc<dyn Class>>
+}
+
+impl ClassHandle {
+    /// Constructor.
+    pub fn new<T: Class + 'static>(class: T) -> ClassHandle {
+        let wrapped = Rc::new(class);
+        ClassHandle{class:Some(wrapped)}
+    }
+
+    /// Return the inner class.
+    pub fn get_class(&self) -> Option<Rc<dyn Class>> {
+        self.class.clone()
+    }
+}
+
+impl CloneRef for ClassHandle {}
+
+
+
+// ================================
+// === Native Constructor Class ===
+// ================================
+
+/// Type alias for a function that can create a `Visualization`.
+pub trait VisualizationConstructor = Fn(&Scene) -> InstantiationResult;
+
+#[derive(CloneRef,Clone,Derivative)]
+#[derivative(Debug)]
+#[allow(missing_docs)]
```

Comment:
We are using `allow(missing_docs)` only for fields. Pub structs have to always have docs.